mlx/backend/common/svd.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright © 2023-2024 Apple Inc.
+
+#include "mlx/allocator.h"
+#include "mlx/backend/common/copy.h"
+#include "mlx/primitives.h"
+
+#ifdef ACCELERATE_NEW_LAPACK
+#include <Accelerate/Accelerate.h>
+#else
+#include <lapack.h>
+#endif
+
+namespace mlx::core {
+
+void svd_impl(const array& a, array& u, array& s, array& vt) {
+  // Lapack uses the column-major convention. To avoid having to transpose
+  // the input and then transpose the outputs, we swap the indices/sizes of the
+  // matrices and take advantage of the following identity (see
+  // https://math.stackexchange.com/a/30077)
+  //    A = UΣVᵀ
+  //    Aᵀ = VΣUᵀ
+  // As a result some of the indices/sizes are swapped as noted above.
+
+  // Rows and cols of the original matrix in row-major order.
+  const int M = a.shape(-2);
+  const int N = a.shape(-1);
+  const int K = std::min(M, N);
+
+  // A of shape M x N. The leading dimension is N since lapack receives Aᵀ.
+  const int lda = N;
+  // U of shape M x M. (N x N in lapack).
+  const int ldu = N;
+  // Vᵀ of shape N x N. (M x M in lapack).
+  const int ldvt = M;
+
+  size_t num_matrices = a.size() / (M * N);
+
+  // lapack clobbers the input, so we have to make a copy.
+  array in(a.shape(), float32, nullptr, {});
+  in.set_data(allocator::malloc_or_wait(in.nbytes()));
+  copy_inplace(a, in, CopyType::GeneralGeneral);
```

Comment:
Maybe this should be a vector copy when `a` is row contiguous? That would be a lot faster.

You can check with `a.flags().row_contiguous`

Reply:
Agree on the concept, but I think this should be handled in the framework? If you show me how, I can make the copy function itself smarter so that primitives don't need to worry about the input memory layout.

Reply:
It's a good point. We kind of drew the line in the copy API a little bit lower because we needed some more control over it in a lot of cases. There are some slight differences here and there in terms of needs for the type of copy.

Here is the [copy function](https://github.com/ml-explore/mlx/blob/main/mlx/backend/common/copy.cpp#L288) which dispatches based on copy type.

I think you want to use:

- `CopyType::Vector`  if `in.flags().row_contiguous`
- `CopyType::General` otherwise (not `GeneralGeneral` since the output of the copy is always contiguous.

And you could probably use `copy` instead of `copy_inplace` and skip dealing with the allocation.